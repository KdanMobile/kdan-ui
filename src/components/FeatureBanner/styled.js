import styled from 'styled-components';

export const BannerWrapper = styled.section`
  position: relative;
`;

export const BannerInner = styled.div`
  transition: background 1.2s;
  background: ${({ bg }) => bg};
  height: 100vh;
`;
