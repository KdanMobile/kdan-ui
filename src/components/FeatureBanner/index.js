import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { Waypoint } from 'react-waypoint';

import { BannerWrapper, BannerInner } from './styled';

const FeatureBanner = ({ banners, backgroundColors, offset, children }) => {
  const bannerRef = useRef(null);
  const [activeIndex, setActiveIndex] = useState(0);
  const [isFixed, setFixed] = useState(false);
  const [bg, setBg] = useState([]);

  const checkVisible = (elm) => {
    const rect = elm.getBoundingClientRect();
    const elemTop = rect.top;
    const elemBottom = rect.bottom;
    const elemHeight = rect.height;

    return elemTop <= 0 && elemBottom >= elemHeight / 3;
  };

  const handleScrollEvent = () => {
    if (bannerRef.current && checkVisible(bannerRef.current)) {
      setFixed(true);
    } else {
      setFixed(false);
    }
  };

  useEffect(() => {
    window.addEventListener('scroll', handleScrollEvent);
    return () => {
      window.removeEventListener('scroll', handleScrollEvent);
    };
  }, [isFixed]);

  useEffect(() => {
    if (backgroundColors[activeIndex]) {
      setBg(backgroundColors[activeIndex]);
    }
  }, [backgroundColors, activeIndex]);

  return (
    <BannerWrapper ref={bannerRef}>
      {banners.map((banner, index) => (
        <Waypoint
          key={`${backgroundColors[0]}-${index}`}
          onEnter={() => setActiveIndex(index)}
          bottomOffset={offset}
          topOffset={offset}
        >
          <BannerInner bg={bg}>
            {children({ banner, index, isFixed, activeIndex })}
          </BannerInner>
        </Waypoint>
      ))}
    </BannerWrapper>
  );
};

FeatureBanner.defaultProps = {
  offset: '10%'
};

FeatureBanner.propTypes = {
  /**
    banner 裡面需要資料，最後會返回 component 參數 banner。
   */
  banners: PropTypes.arrayOf(PropTypes.object).isRequired,
  /**
    每個 banner 裡面的顏色。如果每個 banner 都只要一種顏色，只需要在 array 裡面傳一種顏色即可。
   */
  backgroundColors: PropTypes.arrayOf(PropTypes.string).isRequired,
  /**
    距離下個 banner offset
   */
  offset: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

export default FeatureBanner;
