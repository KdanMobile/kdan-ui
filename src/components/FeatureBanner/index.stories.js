import React from 'react';

import FeatureBanner from './';
import FeatureScrollBar from '../FeatureScrollBar';
import Flex from '../Flex';

export default {
  component: FeatureBanner,
  title: 'FeatureBanner',
  parameters: {
    docs: {
      description: {
        component: `視覺滾動差 Banner。  
          UI 參考: [Kdan PDF Reader](https://www.kdanmobile.com/zh-tw/pdf-reader)`
      }
    }
  }
};

export const Base = (args) => (
  <FeatureBanner {...args}>{() => <div />}</FeatureBanner>
);
Base.args = {
  backgroundColors: ['#f5a623', '#fe8967', '#67b2fe'],
  banners: [{}, {}, {}]
};

export const WithScrollBar = () => {
  const backgroundColors = ['#f5a623', '#fe8967', '#67b2fe'];
  const banners = [
    {
      title: 'banner'
    },
    {
      title: 'banner'
    },
    {
      title: 'banner'
    }
  ];

  return (
    <FeatureBanner backgroundColors={backgroundColors} banners={banners}>
      {({ banner, index, isFixed, activeIndex }) => (
        <Flex justifyContent='center' alignItems='center' height='100%'>
          <h1>{`${banner.title}-${index + 1}`}</h1>
          <FeatureScrollBar
            top='50%'
            right='150px'
            isFixed={isFixed}
            max={banners.length}
            now={activeIndex + 1}
          />
        </Flex>
      )}
    </FeatureBanner>
  );
};
