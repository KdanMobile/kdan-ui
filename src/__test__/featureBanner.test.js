import React from 'react';
import {
  render,
  cleanup,
  screen,
  fireEvent,
  waitFor
} from '@testing-library/react';
import 'jest-styled-components';
import '@testing-library/jest-dom';

import FeatureBanner from '../components/FeatureBanner';

describe('FeatureBanner', () => {
  afterEach(cleanup);
  const banners = [
    {
      name: 'Apple'
    },
    {
      name: 'Banana'
    }
  ];
  const backgroundColors = ['#000'];

  test('basic', () => {
    const { container } = render(
      <FeatureBanner banners={banners} backgroundColors={backgroundColors}>
        {({ banner }) => <div>{banner.name}</div>}
      </FeatureBanner>
    );
    expect(container).toMatchSnapshot();
  });

  test('should render correct banners data', () => {
    render(
      <FeatureBanner banners={banners} backgroundColors={backgroundColors}>
        {({ banner }) => <div data-testid='banner'>{banner.name}</div>}
      </FeatureBanner>
    );

    expect(screen.getAllByTestId('banner').length).toBe(2);
    screen.getAllByTestId('banner').forEach((element, index) => {
      expect(element.textContent).toBe(banners[index].name);
    });
  });

  test('should render correct banners color if colors is only one', async () => {
    const { container } = render(
      <FeatureBanner banners={banners} backgroundColors={backgroundColors}>
        {() => <div />}
      </FeatureBanner>
    );

    const wrapper = container.querySelectorAll('.sc-AxirZ');
    expect(wrapper[0]).toHaveStyleRule('background', '#000');
    expect(wrapper[1]).toHaveStyleRule('background', '#000');
  });

  test('make sure the isFixed parameter is triggered by scrolling', async () => {
    render(
      <FeatureBanner banners={banners} backgroundColors={backgroundColors}>
        {({ isFixed }) => (
          <div data-testid='banner'>{isFixed ? 'isFixed' : ''}</div>
        )}
      </FeatureBanner>
    );

    fireEvent.scroll(window, { target: { scrollY: 1000 } });
    await waitFor(() => {
      screen.getAllByTestId('banner').forEach((element) => {
        expect(element).toBeInTheDocument();
      });
    });
  });
});
