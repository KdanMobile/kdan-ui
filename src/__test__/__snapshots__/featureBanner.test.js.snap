// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureBanner basic 1`] = `
.c0 {
  position: relative;
}

.c1 {
  -webkit-transition: background 1.2s;
  transition: background 1.2s;
  background: #000;
  height: 100vh;
}

<div>
  <section
    class="c0"
  >
    <div
      class="c1"
    >
      <div>
        Apple
      </div>
    </div>
    <div
      class="c1"
    >
      <div>
        Banana
      </div>
    </div>
  </section>
</div>
`;
